tree_unique_args/src/type_analysis.egg

```diff
@@ -0,0 +1,231 @@
+      )
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Sub x y)))
+      (
+       (HasTypeDemand x)
+       (HasTypeDemand y)
+      )
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Mul x y)))
+      (
+       (HasTypeDemand x)
+       (HasTypeDemand y)
+      )
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (LessThan x y)))
+      (
+       (HasTypeDemand x)
+       (HasTypeDemand y)
+      )
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (And x y)))
+      (
+       (HasTypeDemand x)
+       (HasTypeDemand y)
+      )
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Or x y)))
+      (
+       (HasTypeDemand x)
+       (HasTypeDemand y)
+      )
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Not x)))
+      ((HasTypeDemand x))
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Get e idx)))
+      ((HasTypeDemand e))
+      :ruleset type-analysis)
+
+; Pure Op Compute
+(rule (
+       (HasTypeDemand (Add x y))
+       (HasType x (IntT))
+       (HasType y (IntT))
+      )
+      (
+       (HasType (Add x y) (IntT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Sub x y))
+       (HasType x (IntT))
+       (HasType y (IntT))
+      )
+      (
+       (HasType (Sub x y) (IntT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Mul x y))
+       (HasType x (IntT))
+       (HasType y (IntT))
+      )
+      (
+       (HasType (Mul x y) (IntT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (LessThan x y))
+       (HasType x (IntT))
+       (HasType y (IntT))
+      )
+      (
+       (HasType (LessThan x y) (BoolT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (And x y))
+       (HasType x (BoolT))
+       (HasType y (BoolT))
+      )
+      (
+       (HasType (And x y) (BoolT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Or x y))
+       (HasType x (BoolT))
+       (HasType y (BoolT))
+      )
+      (
+       (HasType (Or x y) (BoolT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Not x))
+       (HasType x (BoolT))
+      )
+      (
+       (HasType (Not x) (BoolT))
+      )
+      :ruleset type-analysis)
+(rule (
+       (HasTypeDemand (Get e n))
+       (HasType e (TupleT tylist))
+      )
+      ((HasType (Get e n) (TypeList-ith tylist n)))
+      :ruleset type-analysis)
+
+; Effectful Ops
+(rule ((HasTypeDemand (Print e)))
+      ((HasType (Print e) (UnitT)))
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Write addr val)))
+      ((HasType (Write addr val) (UnitT)))
+      :ruleset type-analysis)
+(rule ((HasTypeDemand (Read addr)))
+      ((HasType (Read addr) (IntT))) ; all reads are ints
+      :ruleset type-analysis)
+
+; Switch 
+; peel off a branch and demand type
+(rule ((HasTypeDemand (Switch cond (Cons branch rest))))
+      (
+       (HasTypeDemand branch)
+       (HasTypeDemand (Switch cond rest))
```

Comment:
This is a cool solution but I think it has a problem
This creates a `Switch` which has the wrong number of branches compared to its condition. I think we don't want to do that (it would create a term that doesn't typecheck)

Instead, I propose doing a `ListHasUniformTypeDemand` and a corresponding `ListHasUniformType` relation. `ListHasUniformType` would be set to a type if all the elements in the list have the same type, which is what you need to typecheck a `Switch`.